go/vt/vtgate/vstream_manager.go

```diff
@@ -683,6 +691,26 @@ func (vs *vstream) streamFromTablet(ctx context.Context, sgtid *binlogdatapb.Sha
 	}
 }
 
+// isRetriable determines whether we should exit immediately or retry the vstream.
+// The first return value determines if the error is retriable, the second indicates whether
+// the tablet on which the error occurred should be ommitted from the candidate list of tablets
+// to choose from on the retry.
```

Comment:
> When do we stop retrying so that the vtgate VStream client knows that there was an issue and can decide what to do in response? 

The vstream tries up to 3 times before exiting with an error. Although, I missed the piece where we sleep if no candidate is found. I think instead of adding a new context with a hardcoded timeout, it might be better to have an upper limit on how many times the tablet picker will sleep and retry as well. What do you think? 

> So now what was a timeout for the stream — which is likely very long, potentially even a background context — becomes a timeout for TabletPicker retries. And in the event that we've now told it to ignore all tablets... it's going to loop forever and make no progress.

This is a problem even in the current world without these changes right? If no candidates are found, the tablet picker will just retry forever...which is why I think adding a cutoff here is better for all cases.

Reply:
Added `Code_NOT_FOUND` for versions that don't have this fix: https://github.com/vitessio/vitess/pull/13582 which I'm assuming is everything before v18? Reason being that we may get this error `Status: Status{code=NOT_FOUND, description=tablet: cell:<cell> uid:<id> is either down or nonexistent, cause=null` [here](https://github.com/slackhq/vitess/blob/slack-vitess-r14.0.5/go/vt/vtgate/vstream_manager.go#L543) as a result of [this](https://github.com/slackhq/vitess/blob/slack-vitess-r14.0.5/go/vt/discovery/healthcheck.go#L781) call.

Reply:
Spoke in channel and I've removed the `Code_NOT_FOUND` error from being retried. Added a shorter context with a 90s timeout to be used to create the `TabletPicker` in addition to `PickForStreaming`. Chose 90s instead of 60s because `GetTabletPickerRetryDelay()` is 30s itself.